[pigeon] migrate to null-safety

Goddchen opened the pull request:
Migrating pigeon package to null-safety.

I marked this as a draft because I have trouble getting one test to pass: `packages/pigeon/mock_handler_tester/test/widget_test.dart:calling methods with null` and I just can't work it out :/ The call throws an AssertionError, which is the expected behavior. Yet the test fails stating that this error was thrown. Can anyone help?

Goddchen commented:
@googlebot I signed it!

mitchross commented:
I really need this pacakge too... How do I include this fork in my own project?

mitchross commented:
@Goddchen here is your error 
![image](https://user-images.githubusercontent.com/6330506/110492455-3408cc00-80c0-11eb-84fd-1b866751c1e2.png)

Goddchen commented:
thanks @mitchross . I didn't event notice this test failing. It's fixed. But the other one that I mentioned in the PR descriptions still puzzles me.

Goddchen commented:
@mitchross  Yeah, I see that piece of code. What puzzles me really is that if I replace the throwing `.send(null)` code with just an `assert(false);` failing assertion. The whole test just runs fine. What's the difference between my `assert(...)` and the one from the code you just showed? 🤔

mitchross commented:
> @mitchross Yeah, I see that piece of code. What puzzles me really is that if I replace the throwing `.send(null)` code with just an `assert(false);` failing assertion. The whole test just runs fine. What's the difference between my `assert(...)` and the one from the code you just showed? 🤔

Do what you suggested, I think this is a bad test setup, with null-safety, it doesn't even make sense anymore. 

plus with this todo, id go with your suggestion 
![image](https://user-images.githubusercontent.com/6330506/110537995-75fe3600-80f1-11eb-93ea-1a14f328636c.png)

Goddchen commented:
Fixed it! After giving it a bit more thought, my finding was that the generated code was actually not doing exactly what it was intended to do. It should handle the `null` case properly - which it does now.

Goddchen commented:
> Also the `local_tests CHANNEL:master` check fails with the message `No dart files found at: /private/var/folders/sr/b58hwhtj0jbcf4r09zmg0wlc0000gn/T/cirrus-ci-build/packages/pigeon/lib`. Well... Obviously there are Dart files in that folder?! 🤷‍♂️

I still don't get what is wrong with this check... The files are there... What's the problem...

mitchross commented:
> > Also the `local_tests CHANNEL:master` check fails with the message `No dart files found at: /private/var/folders/sr/b58hwhtj0jbcf4r09zmg0wlc0000gn/T/cirrus-ci-build/packages/pigeon/lib`. Well... Obviously there are Dart files in that folder?! 🤷‍♂️
> 
> I still don't get what is wrong with this check... The files are there... What's the problem...

The all checks have passed by github, I think you are fine...

Can you also explain the fix a bit more? I went deep into this code and I'm very curious!

gaaclarke commented on packages/pigeon/lib/java_generator.dart:
```diff
@@ -237,15 +243,15 @@ String _castObject(Field field, List<Class> classes, String varName) {
 /// Generates the ".java" file for the AST represented by [root] to [sink] with the
 /// provided [options].
-void generateJava(JavaOptions options, Root root, StringSink sink) {
+void generateJava(JavaOptions? options, Root root, StringSink sink) {
```
The options should be non-null, className should be non-null, package should be nullable  =)

Goddchen replied:
Alright, `className` is now required.

Goddchen commented:
@gaaclarke Still getting this strange `No dart files found at: /private/var/folders/sr/b58hwhtj0jbcf4r09zmg0wlc0000gn/T/cirrus-ci-build/packages/pigeon/lib` error on `local_tests CHANNEL:master`. Can you make any sense out of it? I find it especially strange since the same thing runs just fine in the `local_tests CHANNEL:stable` check 🤔

gaaclarke commented:
@Goddchen I'm not sure, I thought I saw that before in the past but I can't remember where or why.  FWIW in my recent PR I switched the `dartanalyzer` call to `dart analyze`.  That might fix your issue if you rebase.

Goddchen commented:
Dang, I think it just messed up 🙈 I pulled the master changes instead of rebasing. Can you hint me on how to fix this mess? Sorry 😞

gaaclarke commented:
```
git reset --hard <hash before your merge, d61a774?>
git rebase -i <upstream remote>/master
git push -f <remote for your changes>
```

Goddchen commented:
> ```
> git reset --hard <hash before your merge, d61a774?>
> git rebase -i <upstream remote>/master
> git push -f <remote for your changes>
> ```

I guess it worked 🥳 And I guess from your prompt reply that you see this too often 😉  Thanks

Goddchen commented:
@gaaclarke can you help me out here again? 🙈 Those conflicts came up and I'm trying to resolve them. I do rebase on `upstream remote/master` but then somehow all of my changes from this PR are suddenly gone and I get 21 changed files (+ the 5 conflicts) in VSCode. I can't wrap my head around what's happening...

Goddchen commented:
@gaaclarke We're up-to-date and running again 👍

mitchross commented:
@jmagman could we get this merged?

gaaclarke commented:
@mitchross In order to get it merged Goddchen has to request a code review from someone on the Flutter team (it'd likely be deferred to me).  Then they need to go through the review process.  I just happened to glance at the code the other day and did actually review it.  I'm waiting for @Goddchen to be ready.

Goddchen commented:
I didn't know that I had to do anything. I used the "re-request review" button at the top now, was that correct? 🤔

gaaclarke commented:
Yep, thanks @Goddchen.  This is a big review, I'll try to get to it soon.  I'm balancing a few different priorities.  If a couple weeks pass you won't be bothering me if you ping me.

gaaclarke reviewed (commented):
Hey @Goddchen, looks like you accidentally added some changes for files outside of pigeon.

gaaclarke commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -47,27 +47,28 @@ void _writeHostApi(DartOptions opt, Indent indent, Api api) {
+\tfinal Map<Object$nullTag, Object$nullTag> error =\n\t\t\treplyMap['${Keys.error}']! as Map<Object$nullTag, Object$nullTag>;
 \tthrow PlatformException(
-\t\tcode: error['${Keys.errorCode}'] as String,
+\t\tcode: error['${Keys.errorCode}'] as String? ?? 'unknown',
```
Sorry, this might conflict with this PR: https://github.com/flutter/flutter/issues/78336  I had to try to get that landed to address some customer's issues.

I think I like this solution better than the one I did there.  If you end up with a merge conflict, feel free to do this instead.

gaaclarke commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -140,11 +140,13 @@ void _writeFlutterApi(
                 indent.writeln('// ignore message');
                 call = 'api.${func.name}()';
               } else {
+                if (!opt.isNullSafe) {
```
I don't believe this is correct.  If we are generating code with null safety, message will be an `Object?`, so the null check would still be needed here.  I don't think we have a test that covers this unfortunately.  I think the only test that runs this is non-null-safe.  You can see that here: https://github.com/flutter/packages/blob/344f575216cacfcf09d477cf2a32fbe1fac574d3/packages/pigeon/mock_handler_tester/test/widget_test.dart#L70

Goddchen replied:
Yeah it will be `Object?` in null safe mode. But then we don't need to check for `null` since `null` is valid value. It will be passed to `$argType.decode(Object? message)` and there the `null` message will be handled correctly. Or do I misunderstand something here?

gaaclarke replied:
There might be merit to what you are saying.  The problem is this PR has enough changes already.  I'm putting this back the way it was, including the test and signature for decode.  You can file this in a different PR after if you think it's the right change.

gaaclarke commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -236,9 +238,13 @@ void generateDart(DartOptions opt, Root root, StringSink sink) {
       indent.writeln('');
       indent.write(
-        'static ${klass.name} decode(Object message) ',
+        'static ${klass.name} decode(Object$nullTag message) ',
```
I don't think we should change this to `Object?`.  Do you have a reason why you are doing that?

gaaclarke commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -248,21 +254,26 @@ void generateDart(DartOptions opt, Root root, StringSink sink) {
-              indent.add(
-                'pigeonMap[\'${field.name}\'] != null ? ${field.dataType}.decode(pigeonMap[\'${field.name}\']$unwrapOperator) : null',
-              );
+              indent.addln('pigeonMap[\'${field.name}\'] != null');
+              indent.inc(2);
+              indent.writeln(
+                  '? ${field.dataType}.decode(pigeonMap[\'${field.name}\']$unwrapOperator)');
+              indent.write(': null');
```
Can you put this back the way it was please.  I know this generates code that is more properly formatted but the pigeon code is harder to maintain.  I prefer to have the pigeon code easier to maintain and just ask people to run a formatter on the generated code (as long as it isn't crazy).  Maybe indent.format would be usable here instead?

Goddchen replied:
I mainly did this because otherwise the GitHub Actions kept failing. Should we change the Actions instead to allow for badly formatted code?

gaaclarke replied:
Na, you can run a formatter on the generated code that jives with your CI, or you can turn off CI for generated code.  It's impossible to make everyone's CI settings happy.

I tweaked the Indent.format so that the output is the same without inc and dec, so in this instance it will be like you expected.

gaaclarke commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -248,21 +254,26 @@ void generateDart(DartOptions opt, Root root, StringSink sink) {
-                'pigeonMap[\'${field.name}\'] != null ? ${field.dataType}.decode(pigeonMap[\'${field.name}\']$unwrapOperator) : null',
-              );
+              indent.addln('pigeonMap[\'${field.name}\'] != null');
+              indent.inc(2);
```
This code is already covered in the other comment but I don't like naked `indent.inc` or `indent.dec`, that's what `indent.scoped` should be for to make sure each inc is paired with a dec.  Feel free to tweak that function if you need it to do more.

gaaclarke commented on packages/pigeon/lib/dart_generator.dart:
```diff
@@ -297,12 +308,13 @@ void generateTestDart(
     if (api.location == ApiLocation.host && api.dartHostTestHandler != null) {
       final Api mockApi = Api(
-        name: api.dartHostTestHandler,
+        name: api.dartHostTestHandler ?? 'null',
```
Can you use the bang here instead, we already established that it is not null in the conditional.  I don't want to accidentally get classes with "null" as their name sometime in the future.

gaaclarke replied:
done

gaaclarke commented on packages/pigeon/lib/pigeon_lib.dart:
```diff
@@ -118,37 +122,40 @@ bool _isFlutterApi(ClassMirror apiMirror) {
 class PigeonOptions {
   /// Path to the file which will be processed.
-  String input;
+  String? input;
```
We could make this non-null if we wanted.  The command line parser enforces that it is present.  Feel free to leave it as is and I can update it some other time.

gaaclarke replied:
I reverted this change, sorry.  It had unintended consequences.  When you run pigeon without arguments, it's fine for this to bo null.

gaaclarke commented on packages/pigeon/lib/pigeon_lib.dart:
```diff
@@ -312,10 +326,15 @@ options:
-    opts.javaOptions.package = results['java_package'];
-    opts.dartOptions.isNullSafe = results['dart_null_safety'];
+    opts.javaOptions = JavaOptions(
+      className: 'Pigeon',
```
We should be generating a classname that matches the name of the file that is getting generated (ie Foobar.java makes class Foobar).  That code should be somewhere here.  If we are overwriting it somewhere else we should probably leave classname nullable so we can get an error if it isn't set, instead of generating with some placeholder name.

gaaclarke replied:
Removed the 'Pigeon' literal

gaaclarke commented on packages/pigeon/lib/pigeon_lib.dart:
```diff
@@ -404,14 +424,12 @@ options:
-    if (options.javaOut != null) {
-      options.javaOptions.className = basenameWithoutExtension(options.javaOut);
+      options.objcOptions?.header = basename(options.objcHeaderOut!);
     }
```
Here's the code that creates the className, it looks like it might have been lost.

gaaclarke replied:
Not lost, accounted for.

gaaclarke commented on packages/pigeon/mock_handler_tester/test/message.dart:
```diff
@@ -1,83 +1,90 @@
 // See also: https://pub.dev/packages/pigeon
 // ignore_for_file: public_member_api_docs, non_constant_identifier_names, avoid_as, unused_import
-// @dart = 2.8
+// @dart = 2.12
```
Did you make the default output null-safe now?  I think that's a good idea.

gaaclarke commented on packages/pigeon/mock_handler_tester/test/widget_test.dart:
```diff
@@ -73,28 +73,23 @@ void main() {
-        );
-      }
-      expect(mock.log, <String>['initialize']);
+      final Object? result = await const BasicMessageChannel<Object?>(
+        'dev.flutter.pigeon.Api.search',
+        StandardMessageCodec(),
+      ).send(null);
+      expect(result, isA<Map<Object?, Object?>>());
+      final Map<Object?, Object?>? resultMap = result as Map<Object?, Object?>?;
+      expect(
+          (resultMap?['result'] as Map<Object?, Object?>?)?['result'], isNull);
+      expect(
+          (resultMap?['result'] as Map<Object?, Object?>?)?['error'], isNull);
+      expect(mock.log, <String>['initialize', 'search']);
```
Ahh, I see you changed the behavior too.  This is a breaking change and we should avoid it.  (Although I know there is a weird issue with this code on the stable branch, on master it works fine).

gaaclarke replied:
reverted as noted in above comment

gaaclarke commented on packages/pigeon/pigeons/all_datatypes.dart:
```diff
@@ -5,6 +5,19 @@
 import 'package:pigeon/pigeon.dart';
 
 class Everything {
+  Everything({
+    required this.aBool,
+    required this.anInt,
+    required this.aDouble,
+    required this.aString,
+    required this.aByteArray,
+    required this.a4ByteArray,
+    required this.a8ByteArray,
+    required this.aFloatArray,
+    required this.aList,
+    required this.aMap,
+  });
```
Hmm, this is a problem.  I would prefer that we make the fields nullable.  That makes more sense since it matches the code that will get generated by Dart.  There is a whole different issue for making non-null types in the generated output.   We'll need to reserve what you have here for that.

gaaclarke replied:
Done

gaaclarke reviewed (changes requested):
Okay, this is awesome work so far.  Thanks!  I'll try to summarize the big changes we need:
1) Remove the changes for the files outside of the pigeon directory
1) Switch the pigeon files to have nullable types in the classes
1) Keep an eye on https://github.com/flutter/flutter/issues/78336, you might have to rebase and there will be a few conflicts, sorry

I'll try to refrain from landing more changes to pigeon while we wait for this to land.  What is your schedule like for addressing the feedback?  If you don't have a lot of time, I could just push some changes to this PR if that's fine with you.

Goddchen commented:
> I'll try to refrain from landing more changes to pigeon while we wait for this to land. What is your schedule like for addressing the feedback? If you don't have a lot of time, I could just push some changes to this PR if that's fine with you.

Nice review @gaaclarke ! I'll get over the points one by one in the coming days. But feel free to jump in and help if you find the time - I don't mind!

Goddchen commented:
![image](https://user-images.githubusercontent.com/673051/111800388-304d2480-88cc-11eb-8f46-eb967f87ec7d.png)

I guess those changed files outside the pigeon folder come from there commits. I have to admit: I have no clue how those commit ended in this PR 🤔 Nor do I have a clue on how to remove those? 🤷‍♂️ I guess it has something to do with properly rebasing or something like that? Do you have an idea @gaaclarke ?

gaaclarke commented:
I got rid of the other file edits by squashing your commits and force pushing.  There was something weird going on with the PR.  When I squash commits I don't expect to get conflicts, but I did.  I'll try rebasing it on master next so are ready to merge, the squashing should have made this easier.  Watch out, I'll have to force push again after this.

gaaclarke commented:
(I had to rebase one last time to to resolve conflicts from a PR that just landed, hopefully that is the last time.)

gaaclarke commented:
Ok, getting closer.  I'm going to grab lunch.  I think there is a bit of documentation that has to happen before I publish this as well.  I'll just write that up in a different PR once this lands.  I think I can get it in by EOD.

gaaclarke reviewed (approved):
Nice, thanks for the contribution!  I'll clean up the rest of the documentation and stuff in another PR before I publish it.

One thing that would have been helpful if if you tried to keep the migration to null safety and your changes to the output as separate PRs.  It was a bit of work to detangle them.

Take care dude!
